File: flutter_application/lib/models/furniture.dart

class Furniture {
  final int id;
  final int propertyId;
  final String name;
  final String status; // "Good", "Damaged", "Disposed"
  final double purchasePrice;
  final String? imageUrl;
  final String? note;
  final DateTime? addedDate;
  
  final List<FurnitureLog> logs; 

  Furniture({
    required this.id,
    required this.propertyId,
    required this.name,
    required this.status,
    this.purchasePrice = 0.0,
    this.imageUrl,
    this.note,
    this.addedDate,
    this.logs = const [],
  });

  factory Furniture.fromJson(Map<String, dynamic> json) {
    var logList = json['logs'] as List?;
    List<FurnitureLog> parsedLogs = logList != null
        ? logList.map((i) => FurnitureLog.fromJson(i)).toList()
        : [];

    return Furniture(
      id: json['id'],
      propertyId: json['property_id'],
      name: json['name'] ?? '',
      status: json['status'] ?? 'Good',
      purchasePrice: json['purchase_price'] != null 
          ? (json['purchase_price'] as num).toDouble() 
          : 0.0,
      imageUrl: json['image_url'],
      note: json['note'],
      addedDate: json['added_date'] != null 
          ? DateTime.parse(json['added_date']) 
          : null,
      logs: parsedLogs,
    );
  }

}

class FurnitureLog {
  final int id;
  final String logType; // "Maintenance", "Damage", "Repair", "Dispose"
  final String description;
  final DateTime date;
  final String? imageUrl;

  FurnitureLog({
    required this.id,
    required this.logType,
    required this.description,
    required this.date,
    this.imageUrl,
  });

  factory FurnitureLog.fromJson(Map<String, dynamic> json) {
    return FurnitureLog(
      id: json['id'],
      logType: json['log_type'] ?? 'Maintenance',
      description: json['description'] ?? '',
      date: json['date'] != null 
          ? DateTime.parse(json['date']) 
          : DateTime.now(),
      imageUrl: json['image_url'],
    );
  }

}

File: backend/models/furniture.py
from datetime import datetime
from database import db

class Furniture(db.Model):
    __tablename__ = "furnitures"

    id = db.Column(db.Integer, primary_key=True)
    
    # Link to the Property
    property_id = db.Column(db.Integer, db.ForeignKey("properties.id"), nullable=False)
    
    name = db.Column(db.String(100), nullable=False)
    status = db.Column(db.String(50), default="Good")  # e.g., 'Good', 'Damaged', 'Repaired', 'Disposed'
    purchase_price = db.Column(db.Float, default=0.0)
    image_url = db.Column(db.String(255), nullable=True)
    note = db.Column(db.Text, nullable=True)
    
    added_date = db.Column(db.DateTime, default=datetime.utcnow)

    property = db.relationship("Property", backref=db.backref("furnitures", lazy=True, cascade="all, delete-orphan"))

    @classmethod
    def create(cls, property_id, name, status="Good", purchase_price=0.0, image_url=None, note=None):
        item = cls(
            property_id=property_id,
            name=name,
            status=status,
            purchase_price=purchase_price,
            image_url=image_url,
            note=note
        )
        db.session.add(item)
        db.session.commit()
        return item

    @classmethod
    def find_by_property_id(cls, property_id):
        """Get all furniture for a specific property."""
        return cls.query.filter_by(property_id=property_id).all()

    @classmethod
    def find_by_id(cls, furniture_id):
        return cls.query.get(furniture_id)


class FurnitureLog(db.Model):
    __tablename__ = "furniture_logs"

    id = db.Column(db.Integer, primary_key=True)
    
    # Link to the Furniture Item
    furniture_id = db.Column(db.Integer, db.ForeignKey("furnitures.id"), nullable=False)
    
    log_type = db.Column(db.String(50), default="Maintenance") # e.g., 'Maintenance', 'Damage', 'Repair'
    description = db.Column(db.Text, nullable=False)
    image_url = db.Column(db.String(255), nullable=True) # Photo of damage or receipt
    
    date = db.Column(db.DateTime, default=datetime.utcnow)

    # Relationship to Furniture
    furniture = db.relationship("Furniture", backref=db.backref("logs", lazy=True, cascade="all, delete-orphan"))

    @classmethod
    def create(cls, furniture_id, description, log_type="Maintenance", image_url=None, date=None):
        if date is None:
            date = datetime.utcnow()
            
        log = cls(
            furniture_id=furniture_id,
            description=description,
            log_type=log_type,
            image_url=image_url,
            date=date
        )
        db.session.add(log)
        db.session.commit()
        return log

    @classmethod
    def find_by_furniture_id(cls, furniture_id):
        return cls.query.filter_by(furniture_id=furniture_id).order_by(cls.date.desc()).all()